cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -173,6 +224,153 @@ CUDF_HOST_DEVICE inline constexpr T shift(T const& val, scale_type const& scale)
   return left_shift<Rep, Rad>(val, scale);
 }
 
+/** @brief Extracts the sign, exponent, and integral significand of a floating point number.
+ *
+ * @note This returns all zero's for +/-0 and denormals, and zero's with an exponent of
+ * INT_MIN for the unrepresentable values +/-inf and NaN's.
+ *
+ * @tparam FloatingType Type of input floating point value
+ * @param floating The floating value to extract the components from
+ * @return A tuple containing the sign (bool), exponent (power-of-2), and integral significand.
+ */
+template <typename FloatingType,
+          typename cuda::std::enable_if_t<cuda::std::is_floating_point_v<FloatingType>>* = nullptr>
+CUDF_HOST_DEVICE inline auto extract_components(FloatingType floating)
+{
+  // This function is implemented using integer and bit operations, which is much
+  // faster than using the FPU functions for each component individually. This is especially
+  // true for doubles, as bottlenecking on the FP64 GPU pipeline can cut performance in half.
+
+  // See comments in FloatingComponentConstants about the floating point format.
+  using Constants    = FloatingComponentConstants<FloatingType>;
+  using IntegralType = typename Constants::IntegralType;
+
+  // Convert floating to integer
+  auto const integer_rep = [&]() {
+    IntegralType integer;
+    memcpy(&integer, &floating, sizeof(floating));
+    return integer;
+  }();
+
+  // First extract the exponent bits and handle its special values
+  auto const exponent_bits = integer_rep & Constants::exponent_mask;
+  if (exponent_bits == 0) {
+    // Because of the understood set-bit not stored in the mantissa, it is not possible
+    // to store the value zero directly. Instead both +/-0 and denormals are represented with
+    // the exponent bits set to zero.
+    // Thus it's fastest to just floor (generally unwanted) denormals to zero.
```

Comment:
Question: is it possible to represent denormals in our decimal fixed-point format? If so, it's possible users won't want to flush denorms to zero. Usually flushing denormals to zero is guarded by a compiler command line flag (e.g. `-ftz` in nvcc).